plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -1842,30 +1861,154 @@ public Optional<ConnectorOutputMetadata> finishInsert(
-        if (commitVersion != handle.getReadVersion() + 1) {
-            throw new TransactionConflictException(format("Conflicting concurrent writes found. Expected transaction log version: %s, actual version: %s",
-                    handle.getReadVersion(),
-                    commitVersion - 1));
+        long currentVersion = getMandatoryCurrentVersion(fileSystem, handle.getLocation());
+        if (currentVersion > readVersion.get()) {
+            String transactionLogDirectory = getTransactionLogDir(handle.getLocation());
+            for (long version = readVersion.get() + 1; version <= currentVersion; version++) {
+                List<DeltaLakeTransactionLogEntry> transactionLogEntries;
+                try {
+                    long finalVersion = version;
+                    transactionLogEntries = getEntriesFromJson(version, transactionLogDirectory, fileSystem)
+                            .orElseThrow(() -> new TrinoException(DELTA_LAKE_BAD_DATA, "Delta Lake log entries are missing for version " + finalVersion));
+                }
+                catch (IOException e) {
+                    throw new TrinoException(DELTA_LAKE_FILESYSTEM_ERROR, "Failed to access table metadata", e);
+                }
+                checkForInsertTransactionConflicts(session.getQueryId(), isolationLevel, handle, sourceTableHandles, new DeltaLakeCommitSummary(transactionLogEntries), version, attemptCount);
+            }
+
+            // Avoid re-reading already processed transaction log entries in case of retries
+            readVersion.set(currentVersion);
         }
-        // it is not obvious why we need to persist this readVersion
-        TransactionLogWriter transactionLogWriter = transactionLogWriterFactory.newWriter(session, handle.getLocation());
-        transactionLogWriter.appendCommitInfoEntry(getCommitInfoEntry(session, commitVersion, Instant.now().toEpochMilli(), INSERT_OPERATION, handle.getReadVersion()));
+        long commitVersion = currentVersion + 1;
+        writeTransactionLogForInsertOperation(session, handle, sourceTableHandles, isolationLevel, dataFileInfos, commitVersion, currentVersion);
+        return commitVersion;
+    }
 
-        ColumnMappingMode columnMappingMode = getColumnMappingMode(handle.getMetadataEntry(), handle.getProtocolEntry());
+    private void checkForInsertTransactionConflicts(
+            String queryId,
+            IsolationLevel isolationLevel,
+            DeltaLakeInsertTableHandle insertTableHandle,
+            List<ConnectorTableHandle> sourceTableHandles,
+            DeltaLakeCommitSummary commitSummary,
+            long version,
+            int attemptCount)
+    {
+        checkNoMetadataUpdates(commitSummary);
+        checkNoProtocolUpdates(commitSummary);
+
+        List<TupleDomain<DeltaLakeColumnHandle>> enforcedPartitionConstraints = sourceTableHandles.stream()
+                .filter(sourceTableHandle -> sourceTableHandle instanceof DeltaLakeTableHandle)
```

Comment:
The engine passes only DeltaLakeTableHandle in the current implementation, right?